### PR TITLE
2X-42151 Hyland Email Engagements Script

### PR DESCRIPTION
**Pull request by raaayden**
Updating the hyland email engagement log into much better and optimized way.

[Wise Ticket Link](https://2x.wise-portal.com/request-summary.php?id=42151)

### CODE REVIEW

File: Clients/Hyland/Normalization Script/hyland_email_engagement_log.sql [hunk 1]
```diff
@@ -56,385 +56,340 @@ INSERT INTO `x-marketing.hyland.db_email_engagements_log` (
...
+    country AS _country,
+    "" AS _persona,
+    lead_lifecycle_stage__c AS _lifecycle_stage,
+    leadsourcedetail AS _lead_source_detail,
+    mostrecentleadsource AS _most_recent_lead_source,
+    mostrecentleadsourcedetail AS _most_recent_lead_source_detail,
+    programs.name AS _program_name,
+    programs.channel AS _program_channel
+  FROM `x-marketing.hyland_marketo.leads` marketo
+  LEFT JOIN
```
Comment (2x-imransyafi) [inline Clients/Hyland/Normalization Script/hyland_email_engagement_log.sql:137]: JOIN table should in the same line as the JOIN statement, and the conditions should be below, and indented.
```
LEFT JOIN LI_ads_title
    ON LI_ads._adid = LI_ads_title._adid
    AND LI_ads.xx = LI_ads_title.xx
```

File: Clients/Hyland/Normalization Script/hyland_email_engagement_log.sql [hunk 2]
```diff
@@ -56,385 +56,340 @@ INSERT INTO `x-marketing.hyland.db_email_engagements_log` (
...
+    `x-marketing.hyland_marketo.programs` programs
+  ON
+    marketo.acquisitionprogramid = CAST(programs.id AS STRING)
+  WHERE
```
Comment (2x-imransyafi) [inline Clients/Hyland/Normalization Script/hyland_email_engagement_log.sql:141]: Formatting for filters in the WHERE clause,

```
WHERE xx
  AND xx
```

File: Clients/Hyland/Normalization Script/hyland_email_engagement_log.sql [hunk 3]
```diff
@@ -56,385 +56,340 @@ INSERT INTO `x-marketing.hyland.db_email_engagements_log` (
...
+    ) AS _campaign_sent_date,
+    IF(
+      _assetid IS NOT NULL,
+      'Yes',
+      'No'
+    ) AS _EMEA_campaign,
+    'EMEA' AS _airtable_segment,
+    _sfcampaignid
+  FROM
+    `x-marketing.hyland_mysql.db_airtable_email_emea`
```
Comment (2x-imransyafi) [inline Clients/Hyland/Normalization Script/hyland_email_engagement_log.sql:166]: base table should be in the same line as the FROM statement. 

FROM `x-marketing.hyland_mysql.db_airtable_email_emea`

File: Clients/Hyland/Normalization Script/hyland_email_engagement_log.sql [hunk 4]
```diff
@@ -56,385 +56,340 @@ INSERT INTO `x-marketing.hyland.db_email_engagements_log` (
...
+    ) AS _campaign_sent_date,
+    IF(
+      _assetid IS NOT NULL,
+      'Yes',
+      'No'
+    ) AS _EMEA_campaign,
+    'Customer Marketing' AS _airtable_segment,
+    '' AS _sfcampaignID
+  FROM
+    `x-marketing.hyland_mysql.db_airtable_email_customermarketing`
```
Comment (2x-imransyafi) [inline Clients/Hyland/Normalization Script/hyland_email_engagement_log.sql:184]: base table should be in the same line as the FROM statement.

File: Clients/Hyland/Normalization Script/hyland_email_engagement_log.sql [hunk 5]
```diff
@@ -56,385 +56,340 @@ INSERT INTO `x-marketing.hyland.db_email_engagements_log` (
...
+    ORDER BY activitydate DESC
+  ) = 1
+),
+unique_click AS (
+  SELECT
+    DISTINCT
+    email_click.*
+  FROM email_click
+  JOIN email_open 
+  ON email_open._leadid = email_click._leadid
```
Comment (2x-imransyafi) [inline Clients/Hyland/Normalization Script/hyland_email_engagement_log.sql:265]: JOIN condition need to be indented.

File: Clients/Hyland/Normalization Script/hyland_email_engagement_log.sql [hunk 6]
```diff
@@ -56,385 +56,340 @@ INSERT INTO `x-marketing.hyland.db_email_engagements_log` (
...
+    '' AS _email,
+    activitydate AS _timestamp,
+    'downloaded' AS _engagement,
+    '' AS _description,
+    CAST(leadid AS STRING) AS _leadid
+  FROM `x-marketing.hyland_marketo.activities_fill_out_form`
+  WHERE primary_attribute_value 
+  NOT LIKE '%TEST 2X%'
+  AND primary_attribute_value 
+  NOT LIKE '%Email Unsubscribe Form%'
```
Comment (2x-imransyafi) [inline Clients/Hyland/Normalization Script/hyland_email_engagement_log.sql:316]: filters in where clause should follow this format:

```
WHERE xx
  AND xx
```

in this case:

```
WHERE primary_attribute_value NOT LIKE '%TEST 2X%'
  AND primary_attribute_value NOT LIKE '%Email Unsubscribe Form%'
```

File: Clients/Hyland/Normalization Script/hyland_email_engagement_log.sql [hunk 7]
```diff
@@ -56,385 +56,340 @@ INSERT INTO `x-marketing.hyland.db_email_engagements_log` (
...
+    COALESCE(REGEXP_EXTRACT(_description, r'[?&]utm_source=([^&]+)'), "Email") AS _utm_source,
+    REGEXP_EXTRACT(_description, r'[?&]utm_medium=([^&]+)') AS _utm_medium,
+    REGEXP_EXTRACT(_description, r'[?&]utm_content=([^&]+)') AS _utm_content,
+    prospect_info.*,
+    CAST(airtable_info._campaign_sent_date AS TIMESTAMP) AS _campaign_sent_date,
+    airtable_info._EMEA_campaign,
+    airtable_info._airtable_segment,
+    airtable_info._sfcampaignid
+  FROM 
+    engagements_combined AS engagements
```
Comment (2x-imransyafi) [inline Clients/Hyland/Normalization Script/hyland_email_engagement_log.sql:368]: base table should be in the same line as the FROM statement.

File: Clients/Hyland/Normalization Script/hyland_email_engagement_log.sql [hunk 8]
```diff
@@ -56,385 +56,340 @@ INSERT INTO `x-marketing.hyland.db_email_engagements_log` (
...
+  LEFT JOIN
+    prospect_info
+  ON
+    engagements._leadid = prospect_info._id
+  JOIN
+    airtable_info
+  ON
+    engagements._campaignID = CAST(airtable_info._id AS STRING)
```
Comment (2x-imransyafi) [inline Clients/Hyland/Normalization Script/hyland_email_engagement_log.sql:376]: Line 369 - 376, please follow JOIN formatting:

```
LEFT JOIN LI_ads_title
    ON LI_ads._adid = LI_ads_title._adid
    AND LI_ads.xx = LI_ads_title.xx
```

File: Clients/Hyland/Normalization Script/hyland_email_engagement_log.sql [hunk 9]
```diff
@@ -56,385 +56,340 @@ INSERT INTO `x-marketing.hyland.db_email_engagements_log` (
...
+  sub_region__c AS _sub_region_c,
+  description AS _description,
+  sfcampaign.id AS _sfcampaignID
 FROM _all
-LEFT JOIN `x-marketing.hyland_salesforce.Campaign` sfcampaign ON sfcampaign.id = _all._sfcampaignid
-LEFT JOIN (SELECT name, id FROM `x-marketing.toolsgroup_salesforce.User`) user ON user.id = sfcampaign.ownerid;
+LEFT JOIN `x-marketing.hyland_salesforce.Campaign` sfcampaign 
+ON sfcampaign.id = _all._sfcampaignid
+LEFT JOIN (SELECT name, id FROM `x-marketing.hyland_salesforce.User`) user 
+ON user.id = sfcampaign.ownerid;
```
Comment (2x-imransyafi) [inline Clients/Hyland/Normalization Script/hyland_email_engagement_log.sql:392]: LINE 389-392,  please follow JOIN formatting:

```
LEFT JOIN LI_ads_title
    ON LI_ads._adid = LI_ads_title._adid
    AND LI_ads.xx = LI_ads_title.xx
```

Also possible to remove the subquery in the LEFT JOIN?

File: Clients/Hyland/Normalization Script/hyland_email_engagement_log.sql [hunk 10]
```diff
@@ -443,36 +398,42 @@ LEFT JOIN (SELECT name, id FROM `x-marketing.toolsgroup_salesforce.User`) user O
...
+    startdate AS _campaign_start_date,
+    enddate AS _campaign_end_date,
+    region__c AS _region_c,
+    sub_region__c AS _sub_region_c,
+    description AS _description,
+  FROM `x-marketing.hyland_mysql.db_airtable_email_emea` airtable
+  LEFT JOIN `x-marketing.hyland_salesforce.Campaign` campaign 
+  ON campaign.id = airtable._sfcampaignid
+  LEFT JOIN (SELECT name, id FROM `x-marketing.hyland_salesforce.User`) user 
+  ON user.id = campaign.ownerid
```
Comment (2x-imransyafi) [inline Clients/Hyland/Normalization Script/hyland_email_engagement_log.sql:418]: LINE 415-418,  please follow JOIN formatting:

```
LEFT JOIN LI_ads_title
    ON LI_ads._adid = LI_ads_title._adid
    AND LI_ads.xx = LI_ads_title.xx
```

Also possible to remove the subquery in the LEFT JOIN?

File: Clients/Hyland/Normalization Script/hyland_email_engagement_log.sql [hunk 11]
```diff
@@ -443,36 +398,42 @@ LEFT JOIN (SELECT name, id FROM `x-marketing.toolsgroup_salesforce.User`) user O
...
-SELECT campaign_aggregate.*, email_aggregate.* EXCEPT(_campaignID,_utmcampaign,_sfdccampaignid)
+SELECT 
+  campaign_aggregate.*, 
+  email_aggregate.* EXCEPT(_campaignID,_utm_campaign,_sfdccampaignID)
 FROM campaign_aggregate
-LEFT JOIN email_aggregate ON email_aggregate._sfdccampaignid = campaign_aggregate._sfcampaignid
-WHERE _campaignstartdate IS NOT NULL
\ No newline at end of file
+LEFT JOIN email_aggregate 
+ON email_aggregate._sfdccampaignid = campaign_aggregate._sfcampaignid
```
Comment (2x-imransyafi) [inline Clients/Hyland/Normalization Script/hyland_email_engagement_log.sql:438]: JOIN conditions should be indented.

**Review by 2x-imransyafi — CHANGES_REQUESTED** [review]

**Comment by raaayden** [comment]
This pull request is ready for another review.

**Review by 2x-imransyafi — APPROVED** [review]

**Pull request merged**